### ISSUE TITLE
Implement RDF Dataset Canonicalization and provide EARL report

### ISSUE THREAD

**Issue by gkellogg** [apache/jena#2105]
Please consider submitting an EARL implementation report for [RDF Dataset Canonicalization](https://w3.org/TR/rdf-canon).

The implementation report at https://w3c.github.io/rdf-canon/reports/ for CR transition already has five compatible implementations, but we'd like to see reports from other major implementations, such as Jena. The implementation report includes instructions for formatting and submitting new reports.

The test suite can be found at https://w3c.github.io/rdf-canon/tests/.

**Comment by afs** [comment]
There would need to be an implementation before a report can be sent :smile:

I've changed the issue title to reflect that.

**Comment by filip26** [comment]
FYI here is an implementation of URDNA2015 based on Titanium RDF API. https://github.com/setl/rdf-urdna  It's connected to https://github.com/apache/jena/discussions/2216

**Comment by kishorebanala** [comment]
Hey @afs, are you still looking for contributions for this implementation?

**Comment by rvesse** [comment]
@kishorebanala Yes, nobody else in the community has shown any interest in implementing this

**Comment by afs** [comment]
Hi @kishorebanala - yes, it would be great to have dataset canoicalization.

The work will need to check the state of the n-quads writer to write normalized n-quads as required by https://www.w3.org/TR/rdf-canon/#canonical-quads .

It shouldn't be major changes needed - n-quads isn't that complicated anyway.

The RDF 1.2 N-Quads spec https://www.w3.org/TR/rdf12-n-quads/ should be aligned to the needs of canonicaization - indeed, if its not, then RDF 1.2 N-Quads needs fixing!

There is a test suite: https://github.com/w3c/rdf-tests/tree/main/rdf/rdf12/rdf-n-quads/c14n

**Comment by kishorebanala** [comment]
@afs Thank you for the pointers! 

> The RDF 1.2 N-Quads spec https://www.w3.org/TR/rdf12-n-quads/ should be aligned to the needs of canonicaization - indeed, if its not, then RDF 1.2 N-Quads needs fixing!

It looks like https://www.w3.org/TR/rdf12-n-quads/#canonical-quads and https://www.w3.org/TR/rdf-canon/#canonical-quads are in agreement. How do you suggest I confirm that they're indeed equivalent?

I'm not yet sure of the timelines, but I'll start with n-quads writer and the path towards canonical quads.

**Comment by afs** [comment]
> It looks like https://www.w3.org/TR/rdf12-n-quads/#canonical-quads and https://www.w3.org/TR/rdf-canon/#canonical-quads are in agreement. How do you suggest I confirm that they're indeed equivalent?

Unless there is a glaring anomaly, the RDF 1.2 N-Quads can be seen as the primary text nowadays and the test suite: https://github.com/w3c/rdf-tests/tree/main/rdf/rdf12/rdf-n-quads/c14n can be used - it's quite extensive. The files "*-c14n.nq" are the correct output for the input file of the same name without the "c14n" part.

**Comment by afs** [comment]
https://github.com/apache/jena/pull/3452 adds code to run the c14n tests using the `rdftests` command (which needs a classpath with the test jar artifacts on it).

#3453 is a issue to note the tests need updating.

**Comment by kishorebanala** [comment]
@afs Thanks for setting up the test harness for C14N. 

Given that https://www.w3.org/TR/rdf12-n-quads/#canonical-quads defines the format, and https://www.w3.org/TR/rdf-canon/ an algorithm for canonicalizing datasets with it's own tests - https://w3c.github.io/rdf-canon/tests/, how do you suggest we set this up differently? 

I've created an issue to capture the work for RDFC1.0: https://github.com/apache/jena/issues/3461, please lmk if you have any feedback or suggestions.
